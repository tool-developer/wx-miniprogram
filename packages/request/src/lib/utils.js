import utils from '@tool-developer/wx-utils';
import { RequestError } from './RequestError';

function extend(a, b, thisArg) {
  //
  utils.forEach(b, (val, key) => {
    if (thisArg && typeof val === 'function') {
      //
      a[key] = val.bind(thisArg);
    } else {
      //
      a[key] = val;
    }
  });
  //
  return a;
}

//
export function timeout2Throw(ms, request) {
  return new Promise((_, reject) => {
    const t = setTimeout(() => {
      reject(new RequestError(`timeout of ${ms}ms exceeded`, request));

      clearTimeout(t);
    }, ms);
  });
}

//
export function cancel2Throw({ cancelToken }) {
  return new Promise((_, reject) => {
    //
    if (cancelToken) {
      //
      cancelToken.promise.then(cancel => {
        //
        reject(cancel);
      });
    }
  });
}

export default Object.assign(utils, {
  extend,
});
