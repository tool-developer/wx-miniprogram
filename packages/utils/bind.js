module.exports = require('./miniprogram_dist/bind');