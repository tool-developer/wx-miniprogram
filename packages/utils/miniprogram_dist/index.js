function _typeof(obj) {
  "@babel/helpers - typeof";

  if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
    _typeof = function _typeof(obj) {
      return typeof obj;
    };
  } else {
    _typeof = function _typeof(obj) {
      return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
    };
  }

  return _typeof(obj);
}

var toString = Object.prototype.toString;

function isArray(val) {
  //
  return toString.call(val) === '[object Array]';
} //


function isEmpty(obj) {
  if (!obj) {
    return true;
  } //


  if (obj.length) {
    return false;
  } //


  var keys = Object.keys(obj);

  if (keys.length) {
    return false;
  }

  return true;
}

function isObject(val) {
  return val !== null && _typeof(val) === 'object';
}

function isDate(val) {
  //
  return toString.call(val) === '[object Date]';
}

function forEach(obj, fn) {
  // Don't bother if no value provided
  if (obj === null || typeof obj === 'undefined') {
    //
    return;
  } // Force an array if not already something iterable


  if (!isObject(obj)) {
    /* eslint no-param-reassign:0 */
    obj = [obj];
  }

  if (isArray(obj)) {
    // Iterate over array values
    for (var i = 0, l = obj.length; i < l; i += 1) {
      //
      fn.call(null, obj[i], i, obj);
    }
  } else {
    // Iterate over object keys
    var keys = Object.keys(obj);

    for (var _i = 0; _i < keys.length; _i += 1) {
      var key = keys[_i]; //

      if (Object.prototype.hasOwnProperty.call(obj, key)) {
        //
        fn.call(null, obj[key], key, obj);
      }
    }
  }
}

function merge() {
  //
  var result = {}; //

  function assignValue(val, key) {
    if (isObject(result[key]) && isObject(val)) {
      //
      result[key] = merge(result[key], val);
    } else {
      //
      result[key] = val;
    }
  }

  for (var i = 0, l = arguments.length; i < l; i += 1) {
    //
    forEach(i < 0 || arguments.length <= i ? undefined : arguments[i], assignValue);
  } //


  return result;
}

function deepMerge() {
  //
  var result = isArray(arguments.length <= 0 ? undefined : arguments[0]) ? [] : {}; //

  function assignValue(val, key) {
    //
    if (isObject(result[key]) && isObject(val)) {
      //
      result[key] = deepMerge(result[key], val);
    } else if (isObject(val)) {
      //
      if (isArray(val)) {
        //
        result[key] = deepMerge([], val);
      } else {
        //
        result[key] = deepMerge({}, val);
      }
    } else {
      //
      result[key] = val;
    }
  } //


  for (var i = 0, l = arguments.length; i < l; i += 1) {
    //
    forEach(i < 0 || arguments.length <= i ? undefined : arguments[i], assignValue);
  } //


  return result;
}

var index = {
  forEach: forEach,
  merge: merge,
  deepMerge: deepMerge,
  isArray: isArray,
  isObject: isObject,
  isDate: isDate,
  isEmpty: isEmpty
};

export default index;
export { deepMerge, forEach, isArray, isDate, isEmpty, isObject, merge };
