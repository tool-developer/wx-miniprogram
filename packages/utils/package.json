{
  "name": "@tool-developer/wx-utils",
  "version": "0.0.2-canary.0",
  "description": "wx miniprogram utils library",
  "main": "index.js",
  "scripts": {
    "test": "jest",
    "clean": "../../scripts/clean_package.sh",
    "build": "npm run clean & rollup -c"
  },
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com/tool-developer/wx-miniprogram.git"
  },
  "keywords": [
    "wx",
    "utils",
    "wx utils",
    "wx-utils"
  ],
  "author": "@tool-developer",
  "license": "MIT",
  "miniprogram": "miniprogram_dist",
  "bugs": {
    "url": "https://github.com/tool-developer/wx-miniprogram/issues"
  },
  "homepage": "https://github.com/tool-developer/wx-miniprogram#readme",
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "91796edbe4295502fa05e8551eac4cfd4185047c"
}
