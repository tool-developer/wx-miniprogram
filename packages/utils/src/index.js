const { toString } = Object.prototype;

export function isArray(val) {
  //
  return toString.call(val) === '[object Array]';
}
//
export function isEmpty(obj) {
  if (!obj) {
    return true;
  }
  //
  if (obj.length) {
    return false;
  }
  //
  const keys = Object.keys(obj);
  if (keys.length) {
    return false;
  }

  return true;
}

export function isObject(val) {
  return val !== null && typeof val === 'object';
}

export function isDate(val) {
  //
  return toString.call(val) === '[object Date]';
}

export function forEach(obj, fn) {
  // Don't bother if no value provided
  if (obj === null || typeof obj === 'undefined') {
    //
    return;
  }

  // Force an array if not already something iterable
  if (!isObject(obj)) {
    /* eslint no-param-reassign:0 */
    obj = [obj];
  }

  if (isArray(obj)) {
    // Iterate over array values
    for (let i = 0, l = obj.length; i < l; i += 1) {
      //
      fn.call(null, obj[i], i, obj);
    }
  } else {
    // Iterate over object keys
    const keys = Object.keys(obj);
    for (let i = 0; i < keys.length; i += 1) {
      const key = keys[i];
      //
      if (Object.prototype.hasOwnProperty.call(obj, key)) {
        //
        fn.call(null, obj[key], key, obj);
      }
    }
  }
}

export function merge(...args) {
  //
  const result = {};
  //
  function assignValue(val, key) {
    if (isObject(result[key]) && isObject(val)) {
      //
      result[key] = merge(result[key], val);
    } else {
      //
      result[key] = val;
    }
  }

  for (let i = 0, l = args.length; i < l; i += 1) {
    //
    forEach(args[i], assignValue);
  }
  //
  return result;
}

export function deepMerge(...args) {
  //
  const result = isArray(args[0]) ? [] : {};
  //
  function assignValue(val, key) {
    //
    if (isObject(result[key]) && isObject(val)) {
      //
      result[key] = deepMerge(result[key], val);
    } else if (isObject(val)) {
      //
      if (isArray(val)) {
        //
        result[key] = deepMerge([], val);
      } else {
        //
        result[key] = deepMerge({}, val);
      }
    } else {
      //
      result[key] = val;
    }
  }

  //
  for (let i = 0, l = args.length; i < l; i += 1) {
    //
    forEach(args[i], assignValue);
  }
  //
  return result;
}

export default {
  forEach,
  merge,
  deepMerge,
  isArray,
  isObject,
  isDate,
  isEmpty,
};
