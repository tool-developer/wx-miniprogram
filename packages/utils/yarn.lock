# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


"@tool-developer/wo-base-utils@^0.0.2":
  version "0.0.2"
  resolved "https://registry.npmjs.org/@tool-developer/wo-base-utils/-/wo-base-utils-0.0.2.tgz#7ea76cec2d3fa5d70818aca22fa0abe3c986b64c"
  integrity sha512-Bgec5Cg+gyXQKFgb2fvnJi0nJIsueNvzjtA37FXq164wczJZaONn0pp/HD7zjyNX60YOfBBO1JwbcHEqMyTRcw==
