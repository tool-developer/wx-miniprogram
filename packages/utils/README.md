# @tool-developer/wx-utils
底层公共工具方法。

## 用法
```
npm install --save @tool-developer/wx-utils
```
or
```
yarn add @tool-developer/wx-utils
```

```
import utils from '@tool-developer/wx-utils'

```

## API
### .isEmpty(obj)
是否为空

### .isArray(val)
是否数组

### .isObject(val)
是否对象

### .isDate(val)
是否日期对象

### .merge(...args)
拷贝(合并)操作

### .deepMerge(...args)
深拷贝(合并)操作

### .forEach(obj, fn)
遍历处理
