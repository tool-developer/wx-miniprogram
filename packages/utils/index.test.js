import utils from './index';
//
test('@tool-developer utils test', () => {
  //
  expect(utils).toMatchSnapshot();
  //
  expect(utils.isEmpty([])).toBe(true);
  //
  expect(utils.isEmpty({})).toBe(true);
  //
  expect(utils.isArray(['1', '2'])).toBe(true);
  //
  expect(utils.isObject({})).toBe(true);
  //
  expect(utils.isDate(new Date())).toBe(true);

  let a1 = {
    a: 'a',
  };
  let b1 = {
    b: 'b',
  };
  let c1 = {
    c: 'c',
  };
  //
  expect(utils.merge(a1, b1, c1)).toEqual({
    a: 'a',
    b: 'b',
    c: 'c',
  });

  let a2 = {
    a: {
      a: 'a',
    },
  };
  let b2 = {
    b: {
      b: 'b',
    },
  };
  let c2 = {
    b: {
      c: 'c',
    },
  };
  //
  expect(utils.merge(a2, b2, c2)).toEqual({
    a: {
      a: 'a',
    },
    b: {
      b: 'b',
      c: 'c',
    },
  });

  function mergeConfig(config1 = {}, config2 = {}) {
    const config = {};

    //
    function mergeDeepProperties(val, prop) {
      //
      if (utils.isObject(config2[prop])) {
        //
        config[prop] = utils.deepMerge(config1[prop], config2[prop]);
      } else if (typeof config2[prop] !== 'undefined') {
        //
        config[prop] = config2[prop];
      } else if (utils.isObject(config1[prop])) {
        //
        config[prop] = utils.deepMerge(config1[prop]);
      } else if (typeof config1[prop] !== 'undefined') {
        //
        config[prop] = config1[prop];
      }
    }
    //
    utils.forEach(config1, mergeDeepProperties);
    utils.forEach(config2, mergeDeepProperties);

    //
    return config;
  }

  let deepArrayObjectOne = {
    timeout: 10000,
    headers: { Accept: 'application/json' },
    method: 'get',
    requestType: 'json',
    responseType: 'json',
    baseURL: 'http://192.168.1.91:9000',
    maxCache: 100,
    credentials: 'include',
    mode: 'cors',
  };
  let deepArrayObjectTwo = {
    action: '/a/b/c',
    data: {
      projectId: 1,
      freight: 20,
      groupIds: [],
      imageUrls: [
        '//a/b/c',
      ],
      productDesc: '',
      productLabel: '',
      productName: '1',
      productTitle: '1',
      productType: 1,
      remark: '',
      shareImageUrl: '',
      shareName: '1',
      shipmentType: 2,
      videoUrl: '',
      saleRule: {},
      productId: 79,
    },
    method: 'POST',
    serialized: true,
    key: null,
  };

  let res = mergeConfig(deepArrayObjectOne, deepArrayObjectTwo);
  expect(utils.isArray(res.data.imageUrls)).toEqual(true);
});
