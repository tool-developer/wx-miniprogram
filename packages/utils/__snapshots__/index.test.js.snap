// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@tool-developer utils test 1`] = `
Object {
  "deepMerge": [Function],
  "forEach": [Function],
  "isArray": [Function],
  "isDate": [Function],
  "isEmpty": [Function],
  "isObject": [Function],
  "merge": [Function],
}
`;
